### PR TITLE
chore: Handles replication specs updates as in SDKv2 and enables most of remaining SDKv2 Acc Tests

### PR DESCRIPTION
**Pull request by EspenAlbert**
## Description

Prevents increasing num_shards from 1 when state doesn't use it

Link to any related issue(s): CLOUDP-287995

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals I have added appropriate changelog entries.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/common/update/patch_payload.go [hunk 1]
```diff
@@ -12,28 +12,49 @@ import (
...
+				return true
+			}
+		}
+	}
 	return false
 }
 
 func newAttrPatchOperations(patch jsondiff.Patch, options []PatchOptions) *attrPatchOperations {
-	ignoreInState := []string{}
+	ignoreSuffixInState := []string{}
```
Comment (lantoli) [inline internal/common/update/patch_payload.go]: alternatively you can just initialize them to nil instead of non-nil empty, and append will work fine as well:
```
var (
  ignoreSuffixInState []string
  ignorePrefixInState []string
  includeSuffixInState []string
)
```

File: internal/service/advancedcluster/resource_advanced_cluster_test.go [hunk 2]
```diff
@@ -119,11 +119,6 @@ func TestAccClusterAdvancedCluster_singleShardedMultiCloud(t *testing.T) {
 
 func singleShardedMultiCloudTestCase(t *testing.T, isAcc bool) resource.TestCase {
 	t.Helper()
-	// TODO: Already prepared for TPF but getting this error:
-	// resource_advanced_cluster_test.go:119: Step 1/3 error: Check failed: Check 9/12 error: mongodbatlas_advanced_cluster.test: Attribute 'replication_specs.0.region_configs.0.electable_specs.0.disk_iops' expected to be set
-	// Check 10/12 error: mongodbatlas_advanced_cluster.test: Attribute 'replication_specs.0.region_configs.0.analytics_specs.0.disk_iops' expected to be set
-	// Check 11/12 error: mongodbatlas_advanced_cluster.test: Attribute 'replication_specs.0.region_configs.1.electable_specs.0.disk_iops' expected to be set
-	acc.SkipIfAdvancedClusterV2Schema(t)
```
Comment (lantoli) [inline internal/service/advancedcluster/resource_advanced_cluster_test.go:188]: nice! in case it fails because my other PR to enable data source checks, i'll fix it

File: internal/testutil/unit/http_mocker.go [hunk 3]
```diff
@@ -38,6 +45,10 @@ func IsReplay() bool {
 	return slices.Contains([]string{"yes", "1", "true"}, strings.ToLower(os.Getenv(EnvNameHTTPMockerReplay)))
 }
 
+func IsDataUpdate() bool {
+	return slices.Contains([]string{"yes", "1", "true"}, strings.ToLower(os.Getenv(EnvNameHTTPMockerDataUpdate)))
```
Comment (lantoli) [inline internal/testutil/unit/http_mocker.go]: consider using strconv.ParseBool, like [here](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/92cd53a42be98e785e2edc8f584fce0d99a38587/internal/config/advanced_cluster_v2_schema.go#L12) to have consistency with bool env. vars, 

it accepts: "1", "t", "T", "true", "TRUE", "True",    not "yes" but it should be fine.

same for other occurences

**Review by lantoli — APPROVED** [review]

File: internal/testutil/unit/http_mocker.go (hunk 3 above)
Comment (EspenAlbert) [reply internal/testutil/unit/http_mocker.go]: Thanks, didn't know about it. Changed in 369ceb09bb52cd29de415a6aded502c9edde9f97
Although, it is a lot more verbose :/

File: internal/common/update/patch_payload.go (hunk 1 above)
Comment (EspenAlbert) [reply internal/common/update/patch_payload.go]: Nice. Done in 1e4c80fc63b2882d35922bd495f57480b6aec8f3

File: internal/testutil/unit/http_mocker.go [hunk 4]
```diff
@@ -28,14 +28,36 @@ type MockHTTPDataConfig struct {
...
+	return c
 }
 
 func IsCapture() bool {
-	return slices.Contains([]string{"yes", "1", "true"}, strings.ToLower(os.Getenv(EnvNameHTTPMockerCapture)))
+	val, err := strconv.ParseBool(os.Getenv(EnvNameHTTPMockerCapture))
+	if err != nil {
+		return false
+	}
+	return val
```
Comment (lantoli) [inline internal/testutil/unit/http_mocker.go]:
```suggestion
	val, _ := strconv.ParseBool(os.Getenv(EnvNameHTTPMockerCapture))
	return val
```

false is the default value if something goes wrong in ParseBool, no need to do that if logic, same for all funcs

Comment (EspenAlbert) [reply internal/testutil/unit/http_mocker.go]: Thanks. Changed in 21721295703853641c79d737eaee871af8c23a4e

File: internal/service/advancedclustertpf/resource_upgrade.go [hunk 5]
```diff
@@ -6,7 +6,7 @@ import (
 )
 
 func getTenantUpgradeRequest(state, patch *admin.ClusterDescription20240805) *admin.LegacyAtlasTenantClusterUpgradeRequest {
-	if patch.ReplicationSpecs == nil {
+	if patch == nil || patch.ReplicationSpecs == nil {
```
Comment (lantoli) [inline internal/service/advancedclustertpf/resource_upgrade.go]: is this equivalent to len(patch.GetReplicationSpecs()) == 0 ?  or just  len(patch.ReplicationSpecs) == 0

File: internal/service/advancedclustertpf/resource.go [hunk 6]
```diff
@@ -102,19 +102,26 @@ func (r *rs) Update(ctx context.Context, req resource.UpdateRequest, resp *resou
...
+		IncludeInStateSuffix: []string{"diskIOPS"},
+	}
+	patchReq, err := update.PatchPayload(stateReq, planReq, patchOptions)
 	if err != nil {
 		diags.AddError("errorPatchPayload", err.Error())
 		return
 	}
 	var clusterResp *admin.ClusterDescription20240805
-	if !update.IsZeroValues(patchReq) {
+	if !update.IsZeroValues(patchReq) || findNumShardsUpdates(ctx, &state, &plan, diags) != nil {
```
Comment (lantoli) [inline internal/service/advancedclustertpf/resource.go]: oh, why is this change necessary, isn't it detected in update?

Comment (EspenAlbert) [reply internal/service/advancedclustertpf/resource.go]: `num_shards` is only part of the legacy schema

File: internal/service/advancedclustertpf/resource_upgrade.go (hunk 5 above)
Comment (EspenAlbert) [reply internal/service/advancedclustertpf/resource_upgrade.go]: it is a pointer, so cannot use length

File: internal/service/advancedclustertpf/resource.go (hunk 6 above)
Comment (EspenAlbert) [reply internal/service/advancedclustertpf/resource.go]: Updated again in 63c538e7bf7580a3c8a13e2e401c15bc9625eeab

Comment (EspenAlbert) [reply internal/service/advancedclustertpf/resource.go]: And again 2c0dbde44253bf246efacab595811d973bde69bb

Comment (lantoli) [reply internal/service/advancedclustertpf/resource.go]: got it, thanks

File: internal/service/advancedclustertpf/resource_upgrade.go (hunk 5 above)
Comment (lantoli) [reply internal/service/advancedclustertpf/resource_upgrade.go]: i think this works: 
```
if len(patch.GetReplicationSpecs()) == 0 {
```

GetReplicationSpecs() supports nil in patch and returns a slice:
```
// GetReplicationSpecs returns the ReplicationSpecs field value if set, zero value otherwise
func (o *ClusterDescription20240805) GetReplicationSpecs() []ReplicationSpec20240805 {
	if o == nil || IsNil(o.ReplicationSpecs) {
		var ret []ReplicationSpec20240805
		return ret
	}
	return *o.ReplicationSpecs
}
```

it's ok to to do len of a nil slice and will return 0

Comment (lantoli) [reply internal/service/advancedclustertpf/resource_upgrade.go]: oh, it's even better to do:
```
if !patch.HasReplicationSpecs() {
```

there is also `GetReplicationSpecsOk` to get the bool and the object.


```
// HasReplicationSpecs returns a boolean if a field has been set.
func (o *ClusterDescription20240805) HasReplicationSpecs() bool {
	if o != nil && !IsNil(o.ReplicationSpecs) {
		return true
	}

	return false
}
```

**Review by lantoli — APPROVED** [review]

**Comment by EspenAlbert** [comment]
Merging. Only `TestAccClusterAdvancedClusterConfig_asymmetricGeoShardedNewSchemaAddingRemovingShard` left to fix, details in that test case. Will follow up immediately.

File: internal/service/advancedclustertpf/resource.go [hunk 7]
```diff
@@ -102,13 +102,28 @@ func (r *rs) Update(ctx context.Context, req resource.UpdateRequest, resp *resou
 	if diags.HasError() {
 		return
 	}
-	stateReq := normalizeFromTFModel(ctx, &state, diags, false)
-	planReq := normalizeFromTFModel(ctx, &plan, diags, false)
+	stateUsingLegacy := usingLegacySchema(ctx, state.ReplicationSpecs, diags)
+	planUsingLegacy := usingLegacySchema(ctx, plan.ReplicationSpecs, diags)
+	if planUsingLegacy && !stateUsingLegacy {
+		diags.AddError("error operation not permitted, nums_shards from 1 -> > 1", fmt.Sprintf("cannot increase num_shards to > 1 under the current configuration. New shards can be defined by adding new replication spec objects; %s", DeprecationOldSchemaAction))
```
Comment (AgustinBettati) [inline internal/service/advancedclustertpf/resource.go:108]: nit: Could be good to reuse same const error message from SDK v2 implementation to ensure consistency and know it corresponds to the same validation

Comment (lantoli) [reply internal/service/advancedclustertpf/resource.go:108]: @AgustinBettati I was just discussing with @EspenAlbert about starting to have a file with custom consts and logic

Comment (EspenAlbert) [reply internal/service/advancedclustertpf/resource.go:108]: Added to https://jira.mongodb.org/browse/CLOUDP-288298
